streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -577,28 +575,35 @@ private void runLoop() {
                     polledRecords = false;
                 }
 
-                // only record poll latency is long poll is required
-                if (longPoll) {
-                    streamsMetrics.pollTimeSensor.record(computeLatency(), timerStartedMs);
-                }
+                streamsMetrics.pollTimeSensor.record(computeLatency(), timerStartedMs);
             }
 
-            // try to process one fetch record from each task via the topology, and also trigger punctuate
+            // try to process several records from each task via the topology, and also trigger punctuate
             // functions if necessary, which may result in more records going through the topology in this loop
-            if (totalNumBuffered > 0 || polledRecords) {
-                totalNumBuffered = 0;
-
+            if (polledRecords) {
                 if (!activeTasks.isEmpty()) {
+                    long totalProcessedEachRound = Long.MAX_VALUE;
+                    long totalProcessed = 0;
+                    // Round-robin scheduling by taking one record from each task repeatedly
+                    // until no task has any records left
+                    while (totalProcessedEachRound != 0) {
+                        totalProcessedEachRound = 0;
+                        for (StreamTask task : activeTasks.values()) {
+                            int numBuffered = task.process();
+                            // we processed one record,
+                            // and more are buffered waiting for the next round
+                            if (numBuffered > 0) {
+                                totalProcessedEachRound++;
+                                totalProcessed++;
+                            }
+                        }
+                    }
+                    // go over the tasks again to punctuate or commit
                     for (StreamTask task : activeTasks.values()) {
-
-                        totalNumBuffered += task.process();
-
                         requiresPoll = requiresPoll || task.requiresPoll();
```

Comment:
Do we need `requiresPoll` anymore? As we are draining the buffer won't this always be true?